### PR TITLE
fix issue: make --device works at privileged mode 

### PR DESCRIPTION
**Pull request by wenlxie**
Signed-off-by: wenlxie <wenlxie@ebay.com>

Please provide the following information:
-->

**- What I did**
fixs #35991
**- How I verify it**
Make the docker binary, test following cases

- case 1
```
docker run -it --privileged=true --device /dev/vdc:/dev/vda busybox sh
```
Then I will get the error message 
```
docker: Error response from daemon: linux runtime spec devices: In privileged mode,Path:/dev/vda In Container should be different with any path on host.
```
- case 2
```
docker run -it --privileged=true --device /dev/vdc:/dev/vdx busybox sh
```
Then I enter the docker, 
I can see the device /dev/vdx
I exit the docker, and there is no /dev/vdx at host

**- Description for the changelog**
Make --device options works at privileged mode

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @kolyshkin PTAL

File: daemon/oci_linux.go [hunk 1]
```diff
@@ -98,6 +98,20 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
 		}
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if _, err := os.Stat(deviceMapping.PathInContainer); os.IsNotExist(err) {
+				d, _, err := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+				if err != nil {
+					return err
+				}
+				devs = append(devs, d...)
+			} else {
+				if err == nil {
```
Comment (kolyshkin) [inline daemon/oci_linux.go]: If a user specified `--privileged --device xxx:xxx`, then `xxx` will anyway be available, so specifying `--device xxx:xxx` is redundant, and it should not cause an error. One option is to just ignore it, the other is to have a warning, something like this maybe:

```golang
 log.Warnf("Specifying a device %s is redundant for privileged mode", deviceMapping.PathOnHost)
```

**Review by kolyshkin — CHANGES_REQUESTED** [review]

File: daemon/oci_linux.go (hunk 1 above)
Comment (thaJeztah) [reply daemon/oci_linux.go]: Was thinking about that; `--device` allows setting `r` (read), `w` (write), and `m` (minor) options; what options are used if the device passed through `--device` overlaps with a device on the host?

Comment (kolyshkin) [reply daemon/oci_linux.go]: This is a valid concern which I totally overlooked. I think in case restrictive permissions (anything other than the most permissive `rwm`) are given, the code should return an error as it can't enforce the required permissions (e.g. read-only).

What do you think?

Comment (thaJeztah) [reply daemon/oci_linux.go]: In case of a conflict, yes. Perhaps a warning is ok, after all, `--privileged` is expected to remove any restrictions, so perhaps lifting the restrictions that were specified to be ignored is ok.

@wenlxie what would your expectations be?

Comment (wenlxie) [reply daemon/oci_linux.go]: > If a user specified --privileged --device xxx:xxx, then xxx will anyway be available, so specifying --device xxx:xxx is redundant, and it should not cause an error. One option is to just ignore it, the other is to have a warning, something like this maybe:

Thanks for the comments, will change the code.

>Was thinking about that; --device allows setting r (read), w (write), and m (mknod) options; what options are used if the device passed through --device overlaps with a device on the host?

@thaJeztah @kolyshkin 
For this case,  I'd like  to ignore the restrictions and respect the `--privileged ` option.
Will add a warning for this case.

**Comment by wenlxie** [comment]
@thaJeztah @kolyshkin
Now all of the tests had passed, so if you don't have more comments or concerns, can you please help to approve this PR? Thanks

**Comment by vdemeester** [comment]
cc @kolyshkin @vieux @thaJeztah @cpuguy83

**Comment by fabiand** [comment]
Any ETA on this?

We are affected by this

**Comment by mtanino** [comment]
Hi,
Still waiting this update. Who can make forward this PR? Thanks.

**Comment by wenlxie** [comment]
@thaJeztah  What do you think of this PR? Any more concerns?

**Comment by cpuguy83** [comment]
What is this fixing exactly? The referenced issue does not exist.

**Comment by cpuguy83** [comment]
This also seems to be very low in the stack. I'd expect this to be validated on container create rather than when the container is starting.

**Comment by wenlxie** [comment]
@cpuguy83  https://github.com/moby/moby/issues/35991

File: daemon/oci_linux.go [hunk 2]
```diff
@@ -98,6 +98,30 @@ func setDevices(s *specs.Spec, c *container.Container) error {
 		for _, d := range hostDevices {
 			devs = append(devs, oci.Device(d))
 		}
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				logrus.Warnf("Specifying a device %s is redundant for privileged mode", deviceMapping.PathOnHost)
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.Warnf("Specify  permission %s to device %s   is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
```
Comment (cpuguy83) [inline daemon/oci_linux.go]: Extra space after `Specify`
Also should be `Specifying`.
These could all use some extra context on the log messages... e.g. `logrus.WithField("container", c.ID)`

File: daemon/oci_linux.go [hunk 3]
```diff
@@ -98,6 +98,30 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
+				}
+				continue
+			}
+			if _, err := os.Stat(deviceMapping.PathInContainer); os.IsNotExist(err) {
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.Warnf("Specify  permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
```
Comment (cpuguy83) [inline daemon/oci_linux.go]: Extra space after `Specify`, and should be `Specifying`

File: daemon/oci_linux.go [hunk 4]
```diff
@@ -98,6 +98,30 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
+				}
+				d, _, err := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+				if err != nil {
+					return err
+				}
+				devs = append(devs, d...)
+			} else {
+				if err == nil {
+					return fmt.Errorf("In privileged mode,Path:%s in Container should be different with any path on host ", deviceMapping.PathInContainer)
```
Comment (cpuguy83) [inline daemon/oci_linux.go]: `return errdefs.InvalidParameter(errors.New(errorMsg))`

Additionally this should be validated on container create.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Comment by wenlxie** [comment]
@cpuguy83 code changes as your commnets , PTAL

File: daemon/container.go [hunk 5]
```diff
@@ -291,6 +291,22 @@ func (daemon *Daemon) verifyContainerSettings(platform string, hostConfig *conta
...
+	if hostConfig.Privileged {
+		for _, deviceMapping := range hostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				continue
+			}
+			if _, err := os.Stat(deviceMapping.PathInContainer); os.IsNotExist(err) {
+				continue
+			} else {
+				if err == nil {
+					return nil, errors.Errorf("In privileged mode,Path:%s in Container should be different with any path on host ", deviceMapping.PathInContainer)
```
Comment (cpuguy83) [inline daemon/container.go]: Can we clean up some of the grammar here?
Suggestion:
> container device path must be different from any host device path for privileged mode containers

File: daemon/oci_linux.go [hunk 6]
```diff
@@ -98,6 +100,30 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
+					logrus.WithField("container", c.ID).Warnf("Specifying permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+				}
+				d, _, err := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+				if err != nil {
+					return err
+				}
+				devs = append(devs, d...)
+			} else {
+				if err == nil {
+					return errdefs.InvalidParameter(errors.Errorf("In privileged mode,Path:%s in Container should be different with any path on host ", deviceMapping.PathInContainer))
```
Comment (cpuguy83) [inline daemon/oci_linux.go]: Same comment as above re: grammer

File: daemon/container.go [hunk 7]
```diff
@@ -291,6 +291,22 @@ func (daemon *Daemon) verifyContainerSettings(platform string, hostConfig *conta
...
 	}
 
+	if hostConfig.Privileged {
+		for _, deviceMapping := range hostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				continue
+			}
+			if _, err := os.Stat(deviceMapping.PathInContainer); os.IsNotExist(err) {
+				continue
+			} else {
```
Comment (cpuguy83) [inline daemon/container.go]: This might be cleaner like:

```go
if _, err := os.Stat(...); err != nil {
  if os.IsNotExist(err) {
    continue
  }
  return nil, errors.Wrap(err, "error stating device path in container")
}
return nil, errors.Errorf(...)
```

File: daemon/container.go (hunk 5 above)
Comment (wenlxie) [reply daemon/container.go]: Done

File: daemon/oci_linux.go (hunk 6 above)
Comment (wenlxie) [reply daemon/oci_linux.go]: Done

File: daemon/container.go (hunk 7 above)
Comment (wenlxie) [reply daemon/container.go]: Looks better, tks

File: daemon/oci_linux.go [hunk 8]
```diff
@@ -98,6 +100,31 @@ func setDevices(s *specs.Spec, c *container.Container) error {
 		for _, d := range hostDevices {
 			devs = append(devs, oci.Device(d))
 		}
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				logrus.WithField("container", c.ID).Warnf("Specifying a device %s is redundant for privileged mode", deviceMapping.PathOnHost)
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.WithField("container", c.ID).Warnf("Specifying permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
```
Comment (thaJeztah) [inline daemon/oci_linux.go]: I'd prefer this warning to look more like a warning; it's not really "redundant", but will be ignored (so you'd end up with different permissions than you configured);

Something like:

```
running in privileged mode: custom %s permissions for device %s are ignored
```

or

```
custom %s permissions for device %s are ignored in privileged mode
```

File: daemon/oci_linux.go [hunk 9]
```diff
@@ -98,6 +100,31 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
+					d, _, error := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+					if error != nil {
+						return error
+					}
+					devs = append(devs, d...)
+					continue
+				}
+				return errors.Wrap(err, "error stating device path in contianer")
+			}
+			return errdefs.InvalidParameter(errors.Errorf("container device path: %s must be different from any host device path for privileged mode containers", deviceMapping.PathInContainer))
```
Comment (thaJeztah) [inline daemon/oci_linux.go]: Same question as above: in what situation do we end here? 

The error message seems to be if `deviceMapping.PathOnHost == deviceMapping.PathInContainer` (but that's already handled above, and prints a _warning_ ? (perhaps I'm looking wrong, 😅)

File: daemon/oci_linux.go [hunk 10]
```diff
@@ -98,6 +100,31 @@ func setDevices(s *specs.Spec, c *container.Container) error {
 		for _, d := range hostDevices {
 			devs = append(devs, oci.Device(d))
 		}
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				logrus.WithField("container", c.ID).Warnf("Specifying a device %s is redundant for privileged mode", deviceMapping.PathOnHost)
```
Comment (thaJeztah) [inline daemon/oci_linux.go]: Looking at this again, I'm wondering if it's worth logging this, or if it's ok to just skip this warning: if there are no custom options (permissions) set, then the end result is exactly the same (with or without this device), so _perhaps_ it's ok to just ignore this without logging a warning

File: daemon/container.go [hunk 11]
```diff
@@ -291,6 +291,21 @@ func (daemon *Daemon) verifyContainerSettings(platform string, hostConfig *conta
...
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				continue
+			}
+			if _, err := os.Stat(deviceMapping.PathInContainer); err != nil {
+				if os.IsNotExist(err) {
+					continue
+				}
+				return nil, errors.Wrap(err, "error stating device path in contianer")
+			}
+			return nil, errors.Errorf("container device path: %s must be different from any host device path for privileged mode containers", deviceMapping.PathInContainer)
```
Comment (thaJeztah) [inline daemon/container.go]: In what situation do we end here? (looks like we only get here if `deviceMapping.PathOnHost != deviceMapping.PathInContainer` and if `deviceMapping.PathInContainer` exists.

But the _error_ says (if I interpret correctly); it's not allowed to add a device if `deviceMapping.PathOnHost == deviceMapping.PathInContainer` ?

File: daemon/oci_linux.go [hunk 12]
```diff
@@ -98,6 +100,31 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.WithField("container", c.ID).Warnf("Specifying permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+				}
+				continue
+			}
+
+			if _, err := os.Stat(deviceMapping.PathInContainer); err != nil {
+				if os.IsNotExist(err) {
+					if deviceMapping.CgroupPermissions != "rwm" {
+						logrus.WithField("container", c.ID).Warnf("Specifying permission %s to device %s is redundant for privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
```
Comment (thaJeztah) [inline daemon/oci_linux.go]: Same comment here re: error message wording

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/oci_linux.go (hunk 8 above)
Comment (wenlxie) [reply daemon/oci_linux.go]: Done

File: daemon/oci_linux.go (hunk 9 above)
Comment (wenlxie) [reply daemon/oci_linux.go]: Change the logic to make it more clear

File: daemon/oci_linux.go (hunk 12 above)
Comment (wenlxie) [reply daemon/oci_linux.go]: Done

File: daemon/container.go (hunk 11 above)
Comment (wenlxie) [reply daemon/container.go]: Code changed to make it more clear

Comment (wenlxie) [reply daemon/container.go]: @thaJeztah `deviceMapping.PathOnHost == deviceMapping.PathInContainer` has a `continue`

Comment (thaJeztah) [reply daemon/container.go]: Trying to go through the logic:

1. `--device /dev/vdc:/dev/vdc`, means: `deviceMapping.PathOnHost == deviceMapping.PathInContainer`
  - continue (no error)
2. `--device /dev/vdc:/some/path`, and `/some/path` doesn't exist in the container:
  - continue (no error)
3. `--device /dev/vdc:/some/path`, and `/some/path` cannot be read in the container:
  - error: `error stating device path in container`
4. `--device /dev/vdc:/some/path`, and _exists_ in the container:
  - error: `container device path: /some/path must be different from any host device path for privileged mode containers`

So for 4., the device target in the container is not allowed to overlap with an existing path in the container;

- Is this different for non-privileged containers? (I _guess_ not; but we don't seem to be validating that)
- If `deviceMapping.PathOnHost == deviceMapping.PathInContainer`, should we still validate if `PathInContainer` is not an existing path in the container? (could there be situations where a device exists on the host at `/foo/bar`, but the container has a directory (or file) at that location?)
- The error message confuses me, because (IIUC) it's actually about the _container_ path (not the _host_ path), correct?

(again, apologies if I'm misunderstanding :smile:)

Comment (wenlxie) [reply daemon/container.go]: @thaJeztah 
For step 4:
The device target in the container is not allowed to overlap with an existing path on the host but not in the container.
For privileged mode containers , some hostpath maybe mapped into the container, so I do this check here. I admit it is a bit excessive, or we just need to check the `/dev/` path?  What do you think?

**Comment by runcom** [comment]
ping, any update on this? kube is affected by this

File: daemon/container.go [hunk 13]
```diff
@@ -291,6 +291,21 @@ func (daemon *Daemon) verifyContainerSettings(platform string, hostConfig *conta
...
+	if hostConfig.Privileged {
+		for _, deviceMapping := range hostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				continue
+			}
+			if _, err := os.Stat(deviceMapping.PathInContainer); err != nil {
+				if os.IsNotExist(err) {
+					continue
+				}
+				return nil, errors.Wrap(err, "error stating device path in contianer")
```
Comment (kolyshkin) [inline daemon/container.go]: s/contianer/container/

**Review by kolyshkin — CHANGES_REQUESTED** [review]

**Comment by kolyshkin** [comment]
The only concern I have, will docker CLI user see those errors/warnings about device perms being ignored? If not, that looks like a big issue to me -- a user set perms, they are ignored, and the user is not even aware of that.

File: daemon/oci_linux.go [hunk 14]
```diff
@@ -98,6 +100,30 @@ func setDevices(s *specs.Spec, c *container.Container) error {
...
+			if _, err = os.Stat(deviceMapping.PathInContainer); os.IsNotExist(err) {
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.WithField("container", c.ID).Warnf("custom %s permissions for device %s are ignored in privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+				}
+				d, _, error := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+				if error != nil {
+					return error
+				}
+				devs = append(devs, d...)
+			} else {
```
Comment (thaJeztah) [inline daemon/oci_linux.go]: Was looking if we could get rid old all these nested ifs (for example, but extracting the "non-privileged" and "privileged" branches to separate functions), but may be best to leave that refactoring for another PR

**Review by thaJeztah — CHANGES_REQUESTED** [review]
looks like this needs a rebase 🤗

File: daemon/container.go (hunk 13 above)
Comment (wenlxie) [reply daemon/container.go]: Done

File: daemon/oci_linux.go (hunk 1 above)
Comment (wenlxie) [reply daemon/oci_linux.go]: @kolyshkin It will be at the daemon logs and  user will not see it.

Comment (runcom) [reply daemon/oci_linux.go]: you can still print a warning in the cli by inspecting the dev options before they go to the daemon

**Comment by kolyshkin** [comment]
This needs a documentation update (that goes to https://github.com/docker/cli repo I guess) noting that device permissions are ignored when `--privileged` is set. Otherwise LGTM.

**Comment by olljanat** [comment]
@wenlxie this one needs to be rebased

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

**Comment by akhilerm** [comment]
@wenlxie If you are not working on this, can I take it forward.
cc : @thaJeztah @cpuguy83

**Comment by kolyshkin** [comment]
Carried over in https://github.com/moby/moby/pull/40291
